File: Backend/app/routers/companies.py
from fastapi import APIRouter, Depends, HTTPException, status
from pwdlib import PasswordHash
from pwdlib.hashers.bcrypt import BcryptHasher
from sqlalchemy import or_, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models import AppUser, Company, Role
from app.schemas import CompanyRegisterRequest, CompanyRegisterResponse

router = APIRouter(prefix="/api/v1/companies", tags=["companies"])

# Initialize pwdlib PasswordHash with BcryptHasher
password_hash_context = PasswordHash((BcryptHasher(),))


def get_password_hash(password: str) -> str:
    # Safely truncate byte length to 72 bytes before decoding back to string
    truncated_bytes = password.encode("utf-8")[:72]
    safe_password_str = truncated_bytes.decode("utf-8", errors="ignore")
    return password_hash_context.hash(safe_password_str)


@router.post(
    "/register",
    response_model=CompanyRegisterResponse,
    status_code=status.HTTP_201_CREATED,
)
async def register_company(
    payload: CompanyRegisterRequest,
    db: AsyncSession = Depends(get_db),
) -> CompanyRegisterResponse:
    try:
        async with db.begin():
            existing_company = await db.scalar(
                select(Company).where(
                    or_(
                        Company.company_name == payload.company_name,
                        Company.dti_reg_no == payload.dti_reg_no,
                    )
                )
            )
            if existing_company:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Company name or DTI registration number already exists.",
                )

            existing_user = await db.scalar(
                select(AppUser).where(AppUser.email == payload.email)
            )
            if existing_user:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Email already exists.",
                )

            admin_role_id = await db.scalar(
                select(Role.role_id).where(Role.role_name == "Admin")
            )
            if admin_role_id is None:
                admin_role = Role(role_name="Admin")
                db.add(admin_role)
                await db.flush()
                admin_role_id = admin_role.role_id

            if admin_role_id is None:
                raise ValueError("Unable to resolve Admin role_id.")

            company = Company(
                company_name=payload.company_name,
                dti_reg_no=payload.dti_reg_no,
                doe_no=payload.doe_no,
                primary_branch=payload.primary_branch,
                address=payload.address,
            )
            db.add(company)
            await db.flush()

            if company.company_id is None:
                raise ValueError("Unable to create company_id for app_user foreign key.")

            admin_user = AppUser(
                company_id=company.company_id,
                role_id=admin_role_id,
                first_name=payload.first_name,
                last_name=payload.last_name,
                email=payload.email,
                password_hash=get_password_hash(payload.password),
            )
            db.add(admin_user)
            await db.flush()

        return CompanyRegisterResponse(
            company_id=company.company_id,
            company_name=company.company_name,
            user_id=admin_user.user_id,
            role_id=admin_role_id,
            email=admin_user.email,
            message="Company and admin user registered successfully.",
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        ) from e

File: Backend/Routes/companies.js
import express from "express";
import bcrypt from "bcrypt";
import { pool } from "../db.js";

const router = express.Router();

// ---------------------------------------------------------------------------
// POST /api/companies/register
// Creates a new LPG Company profile & primary Admin User profile simultaneously.
// ---------------------------------------------------------------------------

router.post("/register", async (req, res) => {
  const {
    companyName,
    dtiRegNo,
    doeNo,
    primaryBranch,
    address,
    firstName,
    lastName,
    email,
    password,
  } = req.body;

  // 1. Input Validation
  if (
    !companyName ||
    !dtiRegNo ||
    !address ||
    !firstName ||
    !lastName ||
    !email ||
    !password
  ) {
    return res.status(400).json({
      error:
        "Missing required fields. Please provide companyName, dtiRegNo, address, firstName, lastName, email, and password.",
    });
  }

  const client = await pool.connect();

  try {
    await client.query("BEGIN");

    // 2. Check for duplicate company
    const existingCompany = await client.query(
      `SELECT company_id FROM company WHERE company_name = $1 OR dti_reg_no = $2`,
      [companyName, dtiRegNo]
    );

    if (existingCompany.rows.length > 0) {
      await client.query("ROLLBACK");
      return res
        .status(400)
        .json({ error: "Company name or DTI registration number already registered." });
    }

    // 3. Check for duplicate admin email
    const existingUser = await client.query(
      `SELECT user_id FROM app_user WHERE email = $1`,
      [email]
    );

    if (existingUser.rows.length > 0) {
      await client.query("ROLLBACK");
      return res.status(400).json({ error: "Email address is already in use." });
    }

    // 4. Resolve 'Admin' role ID dynamically
    const roleResult = await client.query(
      `SELECT role_id FROM role WHERE LOWER(role_name) = 'admin' LIMIT 1`
    );

    if (roleResult.rows.length === 0) {
      await client.query("ROLLBACK");
      return res.status(500).json({ error: "Admin role is not configured in database." });
    }

    const adminRoleId = roleResult.rows[0].role_id;

    // 5. Insert Company Record
    const companyResult = await client.query(
      `INSERT INTO company (company_name, dti_reg_no, doe_no, primary_branch, address)
       VALUES ($1, $2, $3, $4, $5)
       RETURNING company_id, company_name, dti_reg_no, doe_no, primary_branch, address`,
      [companyName, dtiRegNo, doeNo || null, primaryBranch || null, address]
    );

    const newCompany = companyResult.rows[0];

    // 6. Hash Password & Insert Admin User Profile
    const saltRounds = 10;
    const passwordHash = await bcrypt.hash(password, saltRounds);

    const userResult = await client.query(
      `INSERT INTO app_user (company_id, role_id, first_name, last_name, email, password_hash, status)
       VALUES ($1, $2, $3, $4, $5, $6, 'Active')
       RETURNING user_id, first_name, last_name, email, status, created_at`,
      [
        newCompany.company_id,
        adminRoleId,
        firstName,
        lastName,
        email,
        passwordHash,
      ]
    );

    const newAdmin = userResult.rows[0];

    await client.query("COMMIT");

    res.status(201).json({
      message: "Company & Admin profile registered successfully.",
      company: {
        companyId: newCompany.company_id,
        companyName: newCompany.company_name,
        dtiRegNo: newCompany.dti_reg_no,
        doeNo: newCompany.doe_no,
        primaryBranch: newCompany.primary_branch,
        address: newCompany.address,
      },
      admin: {
        userId: newAdmin.user_id,
        firstName: newAdmin.first_name,
        lastName: newAdmin.last_name,
        email: newAdmin.email,
        status: newAdmin.status,
        createdAt: newAdmin.created_at,
      },
    });
  } catch (err) {
    await client.query("ROLLBACK");
    console.error("POST /api/companies/register failed:", err.message);
    res.status(500).json({ error: "Failed to register company and admin profile." });
  } finally {
    client.release();
  }
});

export default router;